torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -912,3 +923,182 @@ def save_checkpoint(
+        """
+        Load a Distributed checkpoint
+        """
+        checkpoint_path = self._get_latest_intermediate_checkpoint_path()
+
+        if checkpoint_path is None:
+            raise ValueError(
+                "No intermediate checkpoint found in the specified directory. Please ensure that a checkpoint exists."
+            )
+
+        if self._is_rank_zero:
+            logger.info(f"Loading checkpoint from {checkpoint_path}")
+
+        load(
+            state_dict,
+            storage_reader=FileSystemReader(checkpoint_path),
+        )
+
+        return state_dict
+
+    def save_checkpoint(
+        self,
+        state_dict: Dict[str, Any],
+        epoch: int,
+        save_async: bool = False,
+    ) -> None:
+        """
+        Save a distributed checkpoint to disk.
+        If ``save_async`` is True, the save happens asynchronously unblocking the GPUs faster.
+
+        Args:
+            state_dict (Dict[str, Any]): Checkpoint state dict to be written out to file
+            epoch (int): Epoch number. Used to create the checkpoint file name
+            save_async (bool): If True, save the checkpoint asynchronously
+        """
+
+        if self._is_rank_zero:
+            logger.info(
+                f"DistributedCheckpointer is saving a checkpoint for the epoch {epoch}"
+            )
+
+        checkpoint_path = Path.joinpath(
+            self._output_dir, f"{self._checkpoint_dir_prefix}_{epoch}"
+        )
+
+        if self._checkpoint_future is not None:
+            # Previous checkpoint needs to finish before saving the next one.
+            wait_start = time.perf_counter()
+
+            if self._is_rank_zero:
+                logger.info(
+                    "Previous checkpoint has not finished. Checkpointing frequency is too high. Waiting..."
+                )
+
+            self._checkpoint_future.result()
+
+            if self._is_rank_zero:
+                logger.info(
+                    f"Waited {time.perf_counter() - wait_start:.2f} seconds for previous checkpoint to finish"
+                )
+
+        cp_start = time.perf_counter()
+
+        if save_async:
+
+            def callback(
+                f: Future,
+            ) -> None:
+                if f.exception() is None:
+                    if self._is_rank_zero:
+                        logger.info(
+                            f"Checkpoint is saved asynchronously to {checkpoint_path}"
+                        )
+
+                    for index in range(epoch):
+                        last_checkpoint = os.path.join(
+                            self._checkpoint_dir,
+                            f"{self._checkpoint_dir_prefix}_{index}",
+                        )
+                        if os.path.isfile(
+                            os.path.join(last_checkpoint, self._metadata_file)
+                        ):
+                            if self._is_rank_zero:
+                                logger.info(
```

Comment:
do we want to automatically remove the previous checkpoint? some users may want to keep them around so they can evaluate model checkpoints over the course of training

Reply:
Without this, we will be limiting the number of checkpoints we can take as its gonna be bound by the storage available on the host and it can run out quickly for large models. For example, Llama 3.2 vision90B  is almost 150-200GB  in size.
For internal trainers, we keep the last checkpoint only, as for fault tolerance we always restore the latest one. We can possibly make it a config so the users can configure the behavior if they want to evaluate intermediate checkpoints and can also choose to keep only the latest, if the checkpoints are large and taken frequently.

Reply:
Yeah agreed with making it configurable. For this PR I would actually lean towards keeping all intermediate checkpoints around. Mainly because that's what we do for our other checkpointers and we shouldn't have diverging behavior across different checkpoint formats. I'd like to see us add a field like `save_last_n_checkpoints` to all checkpointers in one go

Reply:
Yes. Exactly. It seems like we would need to have a set of checkpointing configs for the user to tune the performance and behavior. We should do it in one go across Checkpointers and get these configs documented. I will get it cleaned up from this PR./

Reply:
@felipemello1 We decided to keep the current behavior consistent with DistributedCheckpointer as well in this PR, which is to keep all the checkpoints persisted for now. I believe when we introduce this feature across Checkpointers, just having `keep_last_n_checkpoints` should be sufficient. 
- None -> All the checkpoints are kept
- 0 -> No intermediate checkpoints
- 1 -> Only the latest valid checkpoint is persisted. (Probably the default)
- N ( N > 1) -> Latest N checkpoints are kept.